File: plugins/index/lang/zig.lua
return function(U)
  local get_text = U.get_text
  local find_child = U.find_child
  local compact_ws = U.compact_ws
  local new_entry = U.new_entry
  local new_import_entry = U.new_import_entry
  local SECTION = U.SECTION
  local CHILD_BRIEF = U.CHILD_BRIEF
  local FIELD_TRUNCATE_THRESHOLD = U.FIELD_TRUNCATE_THRESHOLD
  local truncated_msg = U.truncated_msg

  local function split_path(s)
    local parts = {}
    for p in s:gmatch("[^/]+") do
      parts[#parts + 1] = p
    end
    return parts
  end

  local function get_qualifiers(node, source)
    local parts = {}
    for _, child in ipairs(node:children()) do
      local ck = child:type()
      if ck == "pub" or ck == "export" or ck == "extern"
        or ck == "inline" or ck == "noinline" then
        parts[#parts + 1] = get_text(child, source)
      end
    end
    return table.concat(parts, " ")
  end

  local function fn_sig(node, source)
    local name_node = node:field("name")[1]
    if not name_node then return nil end
    local name = get_text(name_node, source)
    local params = find_child(node, "parameters")
    local params_text = params and get_text(params, source) or "()"
    local ret_node = node:field("type")[1]
    local ret = ret_node and (" " .. get_text(ret_node, source)) or ""
    return compact_ws(name .. params_text .. ret)
  end

  local function find_import_call(node, source)
    for _, child in ipairs(node:children()) do
      if child:type() == "builtin_function" then
        local bi = find_child(child, "builtin_identifier")
        if bi and get_text(bi, source) == "@import" then
          return child
        end
      end
    end
    return nil
  end

  local function find_container(node)
    for _, child in ipairs(node:children()) do
      local ck = child:type()
      if ck == "struct_declaration" or ck == "enum_declaration"
        or ck == "union_declaration" or ck == "opaque_declaration" then
        return child, ck
      end
    end
    return nil, nil
  end

  local function extract_struct_fields(container, source)
    local fields = {}
    local total = 0
    for _, child in ipairs(container:children()) do
      if child:type() == "container_field" then
        total = total + 1
        if total <= FIELD_TRUNCATE_THRESHOLD then
          local name_node = child:field("name")[1]
          local type_node = child:field("type")[1]
          local fname = name_node and get_text(name_node, source) or "_"
          local ftype = type_node and (" " .. get_text(type_node, source)) or ""
          fields[#fields + 1] = compact_ws(fname .. ":" .. ftype)
        end
      end
    end
    if total > FIELD_TRUNCATE_THRESHOLD and #fields < total then
      fields[#fields + 1] = truncated_msg(total)
    end
    return fields
  end

  local function extract_enum_variants(container, source)
    local fields = {}
    local total = 0
    for _, child in ipairs(container:children()) do
      if child:type() == "container_field" then
        total = total + 1
        if total <= FIELD_TRUNCATE_THRESHOLD then
          local name_node = child:field("name")[1]
          fields[#fields + 1] = name_node and get_text(name_node, source) or "_"
        end
      end
    end
    if total > FIELD_TRUNCATE_THRESHOLD and #fields < total then
      fields[#fields + 1] = truncated_msg(total)
    end
    return fields
  end

  local function var_name(node, source)
    local id = find_child(node, "identifier")
    return id and get_text(id, source) or nil
  end

  return {
    import_separator = "/",

    is_doc_comment = function(node, source)
      if node:type() ~= "comment" then return false end
      local text = get_text(node, source)
      return text:sub(1, 3) == "///" and text:sub(1, 4) ~= "////"
    end,

    is_module_doc = function(node, source)
      if node:type() ~= "comment" then return false end
      local text = get_text(node, source)
      return text:sub(1, 3) == "//!"
    end,

    is_test_node = function(node, _source, _attrs)
      return node:type() == "test_declaration"
    end,

    extract_nodes = function(node, source, _attrs)
      local kind = node:type()

      if kind == "function_declaration" then
        local sig = fn_sig(node, source)
        if not sig then return {} end
        local q = get_qualifiers(node, source)
        local text = q ~= "" and (q .. " fn " .. sig) or ("fn " .. sig)
        return { new_entry(SECTION.Function, node, text) }

      elseif kind == "variable_declaration" then
        local name = var_name(node, source)
        if not name then return {} end

        local q = get_qualifiers(node, source)
        local q_prefix = q ~= "" and (q .. " ") or ""

        local is_var = false
        for _, child in ipairs(node:children()) do
          if child:type() == "var" then
            is_var = true
            break
          end
        end
        local kw = is_var and "var " or "const "

        local import_call = find_import_call(node, source)
        if import_call then
          local args = find_child(import_call, "arguments")
          if args then
            local str_node = find_child(args, "string")
            if str_node then
              local raw = get_text(str_node, source)
              local path = raw:match('^"(.-)"$') or raw
              path = path:gsub("%.zig$", "")
              return { new_import_entry(node, { split_path(path) }) }
            end
          end
          return {}
        end

        local container, ckind = find_container(node)
        if container then
          local type_kw = ckind:gsub("_declaration$", "")
          local entry = new_entry(SECTION.Type, node, q_prefix .. type_kw .. " " .. name)
          if ckind == "enum_declaration" then
            entry.children = extract_enum_variants(container, source)
            entry.child_kind = CHILD_BRIEF
          else
            entry.children = extract_struct_fields(container, source)
          end
          return { entry }
        end

        local type_node = node:field("type")[1]
        local type_str = type_node and (": " .. get_text(type_node, source)) or ""
        local text = compact_ws(q_prefix .. kw .. name .. type_str)
        return { new_entry(SECTION.Constant, node, text) }
      end

      return {}
    end,
  }
end

File: maki-code-index/src/index/zig.rs
use tree_sitter::Node;

use super::common::{
    ChildKind, LanguageExtractor, Section, SkeletonEntry, compact_ws, extract_fields_truncated,
    find_child, node_text,
};

pub(crate) struct ZigExtractor;

fn is_type_value_kind(kind: &str) -> bool {
    matches!(
        kind,
        "struct_declaration" | "enum_declaration" | "union_declaration" | "opaque_declaration"
    )
}

impl ZigExtractor {
    fn import_path(bf_node: Node, source: &[u8]) -> Vec<String> {
        let text = find_child(bf_node, "arguments")
            .and_then(|args| find_child(args, "string"))
            .map(|s| node_text(s, source))
            .unwrap_or_default();
        let inner = text.trim_matches('"');
        let clean = inner.strip_suffix(".zig").unwrap_or(inner);
        clean.split('/').map(String::from).collect()
    }

    fn extract_var_decl(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let mut has_pub = false;
        let mut is_const = false;
        let mut name: Option<&str> = None;
        let mut import_node: Option<Node> = None;
        let mut is_type = false;

        let mut cursor = node.walk();
        for child in node.children(&mut cursor) {
            match child.kind() {
                "pub" => has_pub = true,
                "const" => is_const = true,
                "var" | "extern" => {}
                "identifier" if name.is_none() => name = Some(node_text(child, source)),
                "builtin_function" => {
                    if find_child(child, "builtin_identifier")
                        .map(|bi| node_text(bi, source) == "@import")
                        .unwrap_or(false)
                    {
                        import_node = Some(child);
                    }
                }
                k if is_type_value_kind(k) => is_type = true,
                _ => {}
            }
        }

        let name = name?;

        if let Some(bf) = import_node {
            return Some(SkeletonEntry::new_import(
                node,
                vec![Self::import_path(bf, source)],
            ));
        }

        let vis = if has_pub { "pub " } else { "" };
        if is_type {
            let container = Self::find_container(node);
            let type_kw = container
                .as_ref()
                .map(|(_, k)| k.strip_suffix("_declaration").unwrap_or(*k))
                .unwrap_or("const");
            let mut entry =
                SkeletonEntry::new(Section::Type, node, format!("{vis}{type_kw} {name}"));
            if let Some((container_node, container_kind)) = container {
                let is_enum = container_kind == "enum_declaration";
                let fields = extract_fields_truncated(
                    container_node,
                    source,
                    "container_field",
                    |field, src| {
                        let fname = field
                            .child_by_field_name("name")
                            .map(|n| node_text(n, src))
                            .unwrap_or("_");
                        if is_enum {
                            fname.to_string()
                        } else {
                            let ftype = field
                                .child_by_field_name("type")
                                .map(|n| format!(": {}", node_text(n, src)))
                                .unwrap_or_default();
                            format!("{fname}{ftype}")
                        }
                    },
                );
                if is_enum {
                    entry = entry.with_child_kind(ChildKind::Brief);
                }
                entry = entry.with_children(fields);
            }
            return Some(entry);
        }

        let kw = if is_const { "const" } else { "var" };
        let ty = node
            .child_by_field_name("type")
            .map(|n| format!(": {}", node_text(n, source)))
            .unwrap_or_default();
        Some(SkeletonEntry::new(
            Section::Constant,
            node,
            format!("{vis}{kw} {name}{ty}"),
        ))
    }

    fn extract_fn(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let name = node
            .child_by_field_name("name")
            .map(|n| node_text(n, source))?;
        let params = find_child(node, "parameters")
            .map(|n| node_text(n, source))
            .unwrap_or("()");
        let ret = node
            .child_by_field_name("type")
            .map(|n| format!(" {}", node_text(n, source)))
            .unwrap_or_default();

        let mut vis_parts: Vec<&str> = Vec::new();
        let mut cursor = node.walk();
        for child in node.children(&mut cursor) {
            match child.kind() {
                "pub" => vis_parts.push("pub"),
                "inline" => vis_parts.push("inline"),
                "noinline" => vis_parts.push("noinline"),
                "extern" | "export" => vis_parts.push(child.kind()),
                _ => {}
            }
        }

        let vis = if vis_parts.is_empty() {
            String::new()
        } else {
            format!("{} ", vis_parts.join(" "))
        };
        Some(SkeletonEntry::new(
            Section::Function,
            node,
            compact_ws(&format!("{vis}fn {name}{params}{ret}")).into_owned(),
        ))
    }

    fn find_container(node: Node) -> Option<(Node, &'static str)> {
        static KINDS: &[&str] = &[
            "struct_declaration",
            "enum_declaration",
            "union_declaration",
            "opaque_declaration",
        ];
        let mut cursor = node.walk();
        for child in node.children(&mut cursor) {
            if KINDS.contains(&child.kind()) {
                return Some((child, child.kind()));
            }
        }
        None
    }
}

impl LanguageExtractor for ZigExtractor {
    fn extract_nodes(&self, node: Node, source: &[u8], _attrs: &[Node]) -> Vec<SkeletonEntry> {
        let entry = match node.kind() {
            "function_declaration" => self.extract_fn(node, source),
            "variable_declaration" => self.extract_var_decl(node, source),
            _ => None,
        };
        entry.into_iter().collect()
    }

    fn is_doc_comment(&self, node: Node, source: &[u8]) -> bool {
        if node.kind() != "comment" {
            return false;
        }
        let t = node_text(node, source);
        t.starts_with("///") && !t.starts_with("////")
    }

    fn is_module_doc(&self, node: Node, source: &[u8]) -> bool {
        node.kind() == "comment" && node_text(node, source).starts_with("//!")
    }

    fn is_test_node(&self, node: Node, _source: &[u8], _attrs: &[Node]) -> bool {
        node.kind() == "test_declaration"
    }

    fn import_separator(&self) -> &'static str {
        "/"
    }
}
